### PR TITLE
FCE-498 Refactor start device function

### PR DESCRIPTION
**Pull request by kamil-stasiak**
## Description

If you invoke the `start` function without any parameters, it will start the last used device.

## Motivation and Context

It is very confusing to pass `true` as a `deviceId` to the `start` function in order to start the last used device.

## Types of Changes

Refactor

### CODE REVIEW

File: packages/react-client/src/DeviceManager.ts [hunk 1]
```diff
@@ -130,11 +130,12 @@ export class DeviceManager
     this.storageConfig?.saveLastDevice(info);
   }
 
-  // todo `audioDeviceId / videoDeviceId === true` means use last device
-  public async start(deviceId?: string | boolean) {
-    const shouldRestart = !!deviceId && deviceId !== this.deviceState.media?.deviceInfo?.deviceId;
+  public async start(deviceId?: string) {
```
Comment (czerwiukk) [inline packages/react-client/src/DeviceManager.ts:133]: **suggestion:** 
`const useLastDevice = deviceId === undefined;`
In this case `(deviceId || useLastDevice)` is almost always true, unless we pass as an empty string as an argument. (what is rather unwanted i guess)
```
  public async start(deviceId?: string) {
    const shouldRestart = !this.deviceState.media?.deviceInfo?.deviceId;
    const newDevice = deviceId ?? this.getLastDevice()?.deviceId ?? true;
```

File: packages/react-client/src/DeviceManager.ts [hunk 2]
```diff
@@ -130,11 +130,12 @@ export class DeviceManager
...
+    const useLastDevice = deviceId === undefined;
+    const isDeviceStopped = this.deviceState.media?.deviceInfo?.deviceId === undefined;
+    const shouldRestart = (deviceId || useLastDevice) && isDeviceStopped;
 
-    const newDevice = deviceId === true ? this.getLastDevice()?.deviceId || true : deviceId;
+    const newDevice = useLastDevice ? this.getLastDevice()?.deviceId || true : deviceId;
```
Comment (mironiasty) [inline packages/react-client/src/DeviceManager.ts:138]: Why `true` is used here?

File: packages/react-client/src/DeviceManager.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (kamil-stasiak) [reply packages/react-client/src/DeviceManager.ts:133]: If the user does not have a preference for which device to receive, they can start the last selected device (if possible) by invoking `start()`.

If the user wants to start a specific device, they should invoke this function with the device ID: `start(someId)`.

**Comment by kamil-stasiak** [comment]
@mironiasty and @czerwiukk, have a look at the new implementation. The previous one had errors.

File: packages/react-client/src/DeviceManager.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (czerwiukk) [reply packages/react-client/src/DeviceManager.ts:133]: I think simplifying the invocation from `start(undefined)` to `start()` makes sense.

File: packages/react-client/src/DeviceManager.ts [hunk 3]
```diff
@@ -130,22 +130,24 @@ export class DeviceManager
...
-    const exactConstraints = shouldRestart && prepareMediaTrackConstraints(newDevice, trackConstraints);
+    const exactConstraints = shouldProceed && prepareMediaTrackConstraints(newDeviceId, trackConstraints);
     if (!exactConstraints) return;
 
     this.deviceState.mediaStatus = "Requesting";
 
     this.emit(
       "devicesStarted",
-      { ...this.deviceState, restarting: shouldRestart, constraints: newDevice },
+      { ...this.deviceState, restarting: shouldReplaceDevice, constraints: newDeviceId },
```
Comment (mironiasty) [inline packages/react-client/src/DeviceManager.ts:150]: Why we are passing `deviceId` as `constraints`?

File: packages/react-client/src/DeviceManager.ts [hunk 4]
```diff
@@ -130,22 +130,24 @@ export class DeviceManager
...
-    const newDevice = deviceId === true ? this.getLastDevice()?.deviceId || true : deviceId;
+    const shouldReplaceDevice = Boolean(currentDeviceId && currentDeviceId !== newDeviceId);
+    const isDeviceStopped = currentDeviceId === undefined;
+    const shouldProceed = isDeviceStopped || shouldReplaceDevice;
 
     const trackConstraints = this.constraints;
 
-    const exactConstraints = shouldRestart && prepareMediaTrackConstraints(newDevice, trackConstraints);
+    const exactConstraints = shouldProceed && prepareMediaTrackConstraints(newDeviceId, trackConstraints);
     if (!exactConstraints) return;
```
Comment (mironiasty) [inline packages/react-client/src/DeviceManager.ts:144]: I would consider refactoring it to separate function. `deviceId` being `true` for some random reason is... this just should not be written like that.

File: packages/react-client/src/DeviceManager.ts (hunk 3 above)
Comment (kamil-stasiak) [reply packages/react-client/src/DeviceManager.ts]: Good catch, fixed!

File: packages/react-client/src/DeviceManager.ts (hunk 4 above)
Comment (kamil-stasiak) [reply packages/react-client/src/DeviceManager.ts]: You're right, fixed!

**Review by mironiasty — APPROVED** [review]

**Pull request merged**